File: Sources/ImageHub/ViewModels/AppState.swift
import AppKit
import Combine
import SwiftUI

/// Sections in the sidebar. Deliberately not called `Section` — that name
/// belongs to SwiftUI and shadowing it breaks every `Form`.
enum SidebarSection: String, CaseIterable, Identifiable, Hashable {
    case dashboard
    case templates
    case images
    case drives
    case builds

    var id: String { rawValue }

    var label: String {
        switch self {
        case .dashboard: return "Overview"
        case .templates: return "Templates"
        case .images: return "Windows Images"
        case .drives: return "USB Drives"
        case .builds: return "Build History"
        }
    }

    var symbol: String {
        switch self {
        case .dashboard: return "square.grid.2x2"
        case .templates: return "square.stack.3d.up"
        case .images: return "opticaldiscdrive"
        case .drives: return "externaldrive"
        case .builds: return "clock.arrow.circlepath"
        }
    }
}

@MainActor
final class AppState: ObservableObject {
    static let shared = AppState()

    @Published var section: SidebarSection = .dashboard
    @Published var selectedTemplateID: UUID?

    let templates = TemplateStore()
    let library = ImageLibrary()
    let updates = UpdateChecker()

    // Drives
    @Published var drives: [USBDrive] = []
    @Published var rejectedDrives: [RejectedDrive] = []
    @Published var isScanningDrives = false
    @Published var selectedDriveID: String?

    // Builds
    @Published var activeJob: BuildJob?
    @Published var history: [BuildJob] = []
    @Published var showingBuildSheet = false
    /// Pre-selected template when the build sheet is opened from a template row.
    @Published var buildSheetTemplateID: UUID?

    private var driveTimer: AnyCancellable?
    private var cancellables = Set<AnyCancellable>()

    private init() {
        // The stores are nested ObservableObjects. SwiftUI only observes the
        // object a view actually holds, so their change notifications are
        // republished through AppState — otherwise `appState.templates.templates`
        // updates without ever redrawing.
        for publisher in [
            templates.objectWillChange.eraseToAnyPublisher(),
            library.objectWillChange.eraseToAnyPublisher(),
            updates.objectWillChange.eraseToAnyPublisher()
        ] {
            publisher
                .sink { [weak self] _ in
                    // These publishers are owned by @MainActor stores, so this
                    // always runs on the main thread.
                    MainActor.assumeIsolated { self?.objectWillChange.send() }
                }
                .store(in: &cancellables)
        }

        Task { await refreshDrives() }
        startDriveWatch()
    }

    // MARK: - Drives

    /// Drives are polled rather than watched: `DADiskAppeared` callbacks need a
    /// run-loop session and a poll every few seconds is plenty for a picker.
    private func startDriveWatch() {
        driveTimer = Timer.publish(every: 4, on: .main, in: .common)
            .autoconnect()
            .sink { [weak self] _ in
                MainActor.assumeIsolated {
                    guard let self, !self.isScanningDrives else { return }
                    // Don't re-enumerate disks while a build is mid-write.
                    guard self.activeJob?.isRunning != true else { return }
                    Task { await self.refreshDrives() }
                }
            }
    }

    func refreshDrives() async {
        isScanningDrives = true
        let result = await DiskService.scan()
        drives = result.eligible
        rejectedDrives = result.rejected
        if let selected = selectedDriveID, !drives.contains(where: { $0.id == selected }) {
            selectedDriveID = drives.first?.id
        } else if selectedDriveID == nil {
            selectedDriveID = drives.first?.id
        }
        isScanningDrives = false
    }

    var selectedDrive: USBDrive? {
        drives.first { $0.id == selectedDriveID }
    }

    // MARK: - Templates

    var selectedTemplate: DeploymentTemplate? {
        guard let selectedTemplateID else { return nil }
        return templates.template(id: selectedTemplateID)
    }

    func select(_ template: DeploymentTemplate) {
        selectedTemplateID = template.id
        section = .templates
    }

    // MARK: - Builds

    /// Opens the build wizard, optionally pinned to a template.
    func startBuild(template: DeploymentTemplate?) {
        buildSheetTemplateID = template?.id ?? selectedTemplateID ?? templates.templates.first?.id
        showingBuildSheet = true
    }

    /// Kicks off the actual write. Returns the job so the sheet can follow it.
    @discardableResult
    func runBuild(template: DeploymentTemplate, image: WindowsImage, drive: USBDrive) -> BuildJob {
        let job = BuildJob(templateName: template.name, driveName: drive.displayName)
        activeJob = job
        history.insert(job, at: 0)
        if history.count > 25 { history.removeLast(history.count - 25) }

        Task {
            await USBWriter.build(template: template, image: image, drive: drive, job: job)
            if activeJob === job { activeJob = nil }
            await refreshDrives()
        }
        return job
    }

    var isBuilding: Bool { activeJob?.isRunning == true }

    // MARK: - Window

    func openMainWindow() {
        NSApp.activate(ignoringOtherApps: true)
        if let window = NSApp.windows.first(where: { $0.canBecomeMain }) {
            window.makeKeyAndOrderFront(nil)
        }
    }

    // MARK: - Readiness summary for the dashboard

    struct Readiness {
        var hasTemplate: Bool
        var hasImage: Bool
        var hasDrive: Bool
        var hasSplitTool: Bool
        var needsSplitTool: Bool

        var isReady: Bool {
            hasTemplate && hasImage && hasDrive && (!needsSplitTool || hasSplitTool)
        }
    }

    var readiness: Readiness {
        let usableImages = library.images.filter { $0.fileExists }
        return Readiness(
            hasTemplate: templates.templates.contains { $0.isBuildable },
            hasImage: !usableImages.isEmpty,
            hasDrive: !drives.isEmpty,
            hasSplitTool: WimTools.isAvailable,
            needsSplitTool: usableImages.contains { $0.installImageNeedsSplit }
        )
    }
}

File: Windows/App/ViewModels/AppState.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Threading;
using ImageHub.Models;
using ImageHub.Services;
using ImageHub.Support;

namespace ImageHub.ViewModels;

/// <summary>Sections in the navigation pane.</summary>
public enum Section
{
    Dashboard,
    Templates,
    Images,
    Drives,
    Builds,
}

/// <summary>
/// App-wide state: the stores, the drive list, and the build queue.
///
/// The mac app's equivalent republishes its nested stores through Combine; here each
/// store raises its own Changed event and views subscribe to what they display, which
/// is the same idea with less machinery.
/// </summary>
public sealed class AppState : Observable
{
    public static AppState Shared { get; } = new();

    private Section _section = Section.Dashboard;
    private Guid? _selectedTemplateId;
    private string? _selectedDriveId;
    private bool _isScanningDrives;
    private BuildJob? _activeJob;
    private List<UsbDisk> _drives = new();

    /// <summary>
    /// Windows sends a burst of WM_DEVICECHANGE when a stick is plugged in, so the
    /// rescan waits for the burst to finish rather than running a PowerShell query per
    /// message. A DispatcherTimer rather than a Task delay, so the scan and every
    /// notification it causes stay on the UI thread.
    /// </summary>
    private readonly DispatcherTimer _rescanDebounce =
        new() { Interval = TimeSpan.FromMilliseconds(1500) };

    private AppState()
    {
        Templates = new TemplateStore();
        Library = new ImageLibrary();
        Updates = new UpdateChecker();

        _rescanDebounce.Tick += (_, _) =>
        {
            _rescanDebounce.Stop();
            _ = RefreshDrivesAsync();
        };

        Templates.Changed += (_, _) => Raise(nameof(Templates));
        Library.Changed += (_, _) => Raise(nameof(Library));

        SelectedTemplateId = Templates.Templates.FirstOrDefault()?.Id;
    }

    public TemplateStore Templates { get; }

    public ImageLibrary Library { get; }

    public UpdateChecker Updates { get; }

    public Section CurrentSection
    {
        get => _section;
        set
        {
            if (Set(ref _section, value)) { SectionChanged?.Invoke(this, value); }
        }
    }

    public event EventHandler<Section>? SectionChanged;

    // MARK: - Templates

    public Guid? SelectedTemplateId
    {
        get => _selectedTemplateId;
        set
        {
            if (Set(ref _selectedTemplateId, value)) { SelectionChanged?.Invoke(this, EventArgs.Empty); }
        }
    }

    public event EventHandler? SelectionChanged;

    public DeploymentTemplate? SelectedTemplate =>
        SelectedTemplateId is null ? null : Templates.Template(SelectedTemplateId.Value);

    public void Select(DeploymentTemplate template)
    {
        SelectedTemplateId = template.Id;
        CurrentSection = Section.Templates;
    }

    // MARK: - Drives

    public IReadOnlyList<UsbDisk> Drives => _drives;

    public bool IsScanningDrives { get => _isScanningDrives; private set => Set(ref _isScanningDrives, value); }

    public string? SelectedDriveId { get => _selectedDriveId; set => Set(ref _selectedDriveId, value); }

    public UsbDisk? SelectedDrive => _drives.FirstOrDefault(drive => drive.Id == SelectedDriveId);

    public event EventHandler? DrivesChanged;

    public async Task RefreshDrivesAsync()
    {
        if (IsScanningDrives) { return; }
        // Never re-enumerate disks while a build is mid-write.
        if (ActiveJob?.IsRunning == true) { return; }

        IsScanningDrives = true;
        try
        {
            _drives = await DiskService.ScanAsync().ConfigureAwait(true);
            if (SelectedDriveId is null || _drives.All(drive => drive.Id != SelectedDriveId))
            {
                SelectedDriveId = _drives.FirstOrDefault()?.Id;
            }
        }
        finally
        {
            IsScanningDrives = false;
        }
        Raise(nameof(Drives));
        DrivesChanged?.Invoke(this, EventArgs.Empty);
    }

    /// <summary>Called from the window's WM_DEVICECHANGE hook.</summary>
    public void DevicesChanged()
    {
        _rescanDebounce.Stop();
        _rescanDebounce.Start();
    }

    // MARK: - Builds

    public BuildJob? ActiveJob
    {
        get => _activeJob;
        private set
        {
            if (Set(ref _activeJob, value)) { Raise(nameof(IsBuilding)); }
        }
    }

    public List<BuildJob> History { get; } = new();

    public bool IsBuilding => ActiveJob?.IsRunning == true;

    public event EventHandler? HistoryChanged;

    /// <summary>Kicks off the actual write. Returns the job so the dialog can follow it.</summary>
    public BuildJob RunBuild(DeploymentTemplate template, WindowsImage image, UsbDisk drive)
    {
        var job = new BuildJob(template.Name, drive.DisplayName);
        ActiveJob = job;
        History.Insert(0, job);
        if (History.Count > 25) { History.RemoveRange(25, History.Count - 25); }
        HistoryChanged?.Invoke(this, EventArgs.Empty);

        _ = RunBuildAsync(template, image, drive, job);
        return job;
    }

    private async Task RunBuildAsync(
        DeploymentTemplate template,
        WindowsImage image,
        UsbDisk drive,
        BuildJob job)
    {
        // A snapshot, so editing the template while a drive is being written cannot
        // change what lands on it halfway through.
        DeploymentTemplate snapshot = template.DeepCopy();
        await UsbWriter.BuildAsync(snapshot, image, drive, job).ConfigureAwait(true);
        if (ReferenceEquals(ActiveJob, job)) { ActiveJob = null; }
        Raise(nameof(IsBuilding));
        HistoryChanged?.Invoke(this, EventArgs.Empty);
        await RefreshDrivesAsync().ConfigureAwait(true);
    }

    public void ClearFinishedHistory()
    {
        History.RemoveAll(job => !job.IsRunning);
        HistoryChanged?.Invoke(this, EventArgs.Empty);
    }

    // MARK: - Readiness summary for the dashboard

    public sealed class ReadinessReport
    {
        public bool HasTemplate { get; init; }

        public bool HasImage { get; init; }

        public bool HasDrive { get; init; }

        public bool IsElevated { get; init; }

        public bool IsReady => HasTemplate && HasImage && HasDrive && IsElevated;
    }

    public ReadinessReport Readiness => new()
    {
        HasTemplate = Templates.Templates.Any(template => template.IsBuildable),
        HasImage = Library.Images.Any(image => image.FileExists),
        HasDrive = Drives.Count > 0,
        IsElevated = Elevation.IsElevated,
    };
}
